src/Hedgehog/Tree.fs

```diff
@@ -1,120 +1,105 @@
 ﻿namespace Hedgehog
 
-/// A rose tree which represents a random generated outcome, and all the ways
-/// in which it can be made smaller.
 type Tree<'a> =
     | Node of 'a * seq<Tree<'a>>
 
 module Tree =
-    /// The generated outcome.
-    let outcome (Node (x, _) : Tree<'a>) : 'a =
-        x
 
-    /// All the possible shrinks of this outcome. This should be ordered
-    /// smallest to largest as if property still fails with the first shrink in
-    /// the list then we will commit to that path and none of the others will
-    /// be tried (i.e. there is no backtracking).
-    let shrinks (Node (_, xs) : Tree<'a>) : seq<Tree<'a>> =
-        xs
+    let root tree =
+        let (Node (root, _)) = tree
+        root
 
-    /// Create a tree with a single outcome and no shrinks.
-    let singleton (x : 'a) : Tree<'a> =
-        Node (x, Seq.empty)
+    let children tree =
+        let (Node (_, children)) = tree
+        children
 
-    let addChild (child: Tree<'a>) (parent: Tree<'a>) : Tree<'a> =
-        let (Node (x, xs)) = parent
-        Node (x, Seq.cons child xs)
+    let singleton value =
+        Node (value, Seq.empty)
 
-    let addChildValue (a: 'a) (tree: Tree<'a>) : Tree<'a> =
-        tree |> addChild (singleton a)
+    let addChild child parent =
+        Node (root parent, Seq.cons child (children parent))
 
-    let rec cata (f: 'a -> 'b seq -> 'b) (Node (x, xs): Tree<'a>) : 'b =
-        f x (Seq.map (cata f) xs)
+    let addChildValue value parent =
+        parent |> addChild (singleton value)
 
-    let depth (tree: Tree<'a>) : int =
+    let rec cata f tree =
+        f (root tree) (Seq.map (cata f) (children tree))
```

Comment:
I think the pair that is a tree should still be explicitly decomposed.  Does the `.fsi` file prevent this?

Reply:
You can't destructure a parameter with the `.fsi` file and the warning we've enabled. In this case the compiler will call the parameter `_arg1` and we'll get the warning. Also, I'm not a big fan of continually destructuring this way because we already have functions to do this for us.

Reply:
I think you misunderstand the primary purpose of those functions, which is to chain calls.  It is possible (but significantly more difficult...you have to use `Func.flatten`) to reference something twice.  In those cases, I think it is typically more readable to assign that to a local value.  As a bonus, it will also be more efficient.

Not deconstructing directly in the parameters isn't so bad though.  Just do it in the first line of code.

Reply:
> I think you misunderstand the primary purpose of those functions, which is to chain calls. It is possible (but significantly more difficult...you have to use `Func.flatten`) to reference something twice. In those cases, I think it is typically more readable to assign that to a local value. As a bonus, it will also be more efficient.

I think you misunderstand my intentions continuously and it's starting to wear on me, we can talk offline if you'd like.

The purpose of using the calls is to make life easier if the `Tree` representation changes. I'm not in the business of just chaining calls together for the fun of it, or because I can. There is a practical reason to not destructure and I would very much appreciate if that were considered. There is also a middle ground here:

```fsharp
let rec cata f tree =
    let root = root tree
    let children = children tree
    let mapping = cata f
    f root (Seq.map mapping children)
```

I tend to optimize for readability first, and I think you do as well. We have different ideas of what that looks like, but I think we can be a bit more cordial in our approach.

Reply:
> [...] it's starting to wear on me [...]
> [...] I think we can be a bit more cordial in our approach.

Yes, I am sorry. My language was much too harsh there.  I apologize.

> I think you misunderstand my intentions continuously [...]
>
> The purpose of using the calls is to make life easier if the Tree representation changes.  [...]  There is a practical reason to not destructure and I would very much appreciate if that were considered.

Indeed, I misunderstood your intention.  Based on your previous comments, I thought you wanted to restrict this PR to improving function names in `Tree.fs` and adding a `Tree.fsi` file.  I didn't consider that you also wanted to change the implementation to make it easier to change the tree representation.  I think that is a good improvement, I don't mind if such changes are also included in this PR.